Added the TextureStitchEvent.Collected event.

marchermans opened the pull request:
When @liach moved the TextureStitchEvent.Pre call to before the collection of the TextureAtlasSprites of Items and Blocks, in PR #3436, he made it also impossible to register additional Textures that are based on already registered TextureAtlasSprites, as the registered list in the Map is now empty, and there is no way to retrieve an already registered Texture as none have bene registered.

A good example are the materialized greyscale textures of TinkersConstruct, currently maintained by @bonii-xx, or my mod called Armory.

This PR creates a new TextureStitchEvent.Collected that is mend to be fired after the TextureAtlasSprites of the Blocks and Items have been collected the normal way, but before the Stitch actually happens. This makes it possible to register textures based on already registered Textures, before the Stitch happens.

I added a method to ForgeHooksClient to minimalize the changes to the patch in TextureMap,java and added the call to that method on the location of old call to ForgeHooksClient.onTextureStitchPre.


_**UPDATE:**_
It has been pointed out to me that I should explain my reason why i wanted this PR better, as for a simple recolor of an existing texture I could use IItemColor. I agree that this would be the better solution if i just wanted to recolor existing textures with an overlay color. However this is not the only use case for my system that requires this event. I have, next to metalic materials for whom IItemColor whould have been enough, also materials that require an existing texture as the overlay. A good example is Wood. The bark texture of the material is then used as the overlay and retextures the existing greyscale model pixel for pixel. This is not possible by simply using IItemColor as that only allows for individual colors.

_**EXAMPLE:**_
An example of a TextureCreator that is part of my Library SmithsCore: https://github.com/SmithsGaming/SmithsCore/blob/Development-1.11/src/com/smithsmodding/smithscore/client/textures/TextureCreator.java is used to create the individual textures. In this Example a modder can register an Implementation of ITextureController to modify marked textures and reregister them.

marchermans committed:
Added TextureStitchEvent.Collected.

The TextureStitchEvent.Collected is fired after all the textures for Stitching have been collected by the TextureMap, and before the stitch itself.

This allows a Modder to register a modified copy of an already registered Atlas.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Added TextureStitchEvent.Collected.

The TextureStitchEvent.Collected is fired after all the textures for Stitching have been collected by the TextureMap, and before the stitch itself.

This allows a Modder to register a modified copy of an already registered Atlas.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Added TextureStitchEvent.Collected.

The TextureStitchEvent.Collected is fired after all the textures for Stitching have been collected by the TextureMap, and before the stitch itself.

This allows a Modder to register a modified copy of an already registered Atlas.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Added TextureStitchEvent.Collected.

The TextureStitchEvent.Collected is fired after all the textures for Stitching have been collected by the TextureMap, and before the stitch itself.

This allows a Modder to register a modified copy of an already registered Atlas.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3660) <br/>All committers have signed the CLA.

liach commented on src/main/java/net/minecraftforge/client/event/TextureStitchEvent.java:
```diff
@@ -48,6 +48,16 @@ public TextureMap getMap()
+     */
+    public static class Collected extends TextureStitchEvent
+    {
+        public Collected(TextureMap map){ super(map); }
```
Code style

marchermans replied:
What is wrong here? The formatting is the same in the other Events.

liach replied:
Well, yes, all the events are improperly formatted. You can ignore this point, indeed.

liach commented:
> A good example are the materialized greyscale textures of TinkersConstruct, currently maintained by @bonii-xx, or my mod called Armory.

Can you give a link to the code?

marchermans commented:
Added an example and better reasoning why i need the Event.

liach commented:
I still wonder how you use these sprites without references in jsons. That was why I moved this event.

marchermans commented:
The principal system here works in combination with an API and a custom ModelLoader.
The API holds reference to all possible combination that can be made on the Armor in other words:
 - Which materials can be made into the given Base Armor.
 - Which upgrades can be put onto this armor and out of which Materials are they made.

The ModelLoader then loads a JSON which reflects these informations. It holds a base texture for the base Armor, in greyscale, and textures for each upgrade, also in greyscale. (Using events during the loading of the model somebody who is registering additional upgrades for a model can also provide the greyscale textures for that upgrade during loading.) It then returns a unbaked model that has those greyscale textures as its textures.

Before the change those greyscale textures where registered to the Stitcher and I could grab them during the Pre event. Materialize them, so for example color them in a metal color or overlay them with an existing texture of say a wooden bark for a wooden material, and then register those materialized textures to the map so that they would be stitched in properly.

The fact that the baking of the models happens after the textures are loaded allows me to grab the materialized textures during bake and create a new copied model that is retextured with the correct materialized texture during the baking process and store them in a collected list. When handleItemState is called, using the data stored on the stack it will then grab first the correctly textured basemodels quad for the base armor, and then if applicable the quads of all the correctly materialized models of the upgrades and combines them to that it generates one flattened model.

Now that the TextureStitchEvent.Pre is moved to before the collection of the Textures from the Models. It is impossible to grab the greyscale textures that it needs to generate the materialized ones that depend on those. That is the reason why i added the collected event phase into the TextureStitch. At this point the greyscale textures are collected and i can grab them and then retexture those like i did with the Pre event before it was moved.

Greets, Orion

My implementation:
 * Loader: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/loaders/MultiLayeredArmorModelLoader.java
 * UnbakedModel: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/item/unbaked/MultiLayeredArmorItemModel.java
      * UnbakedComponentModel: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/item/unbaked/components/ArmorCoreComponentModel.java
 * BakedModel: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/item/baked/BakedMultiLayeredArmorItemModel.java
      * UnbakedComponentModel: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/item/baked/components/BakedCoreComponentModel.java

marchermans commented:
Should I add a Test Mod for this?

liach commented:
Absolutely!

marchermans committed:
Merge pull request #1 from MinecraftForge/1.11.x

Update to newest 1.11.x Branch.

marchermans committed:
Added a TestMod.

Based on TiCon and Armory Code.

marchermans commented:
I added a test mod and pulled in the latest changes.
Examples of what the test mod does:

![image](https://cloud.githubusercontent.com/assets/5585406/24015714/87a893b0-0a89-11e7-9edb-6839a4626c14.png)

And:

![image](https://cloud.githubusercontent.com/assets/5585406/24015762/ae3755f2-0a89-11e7-83dd-662dd97e880b.png)

marchermans commented:
@liach @RainWarrior Should be good to go.

liach commented:
looks great. Just waiting for fry then.

Madsthunder commented:
@mezz Any idea when the next round of prs will be pulled?

marchermans commented:
@mezz , is there any news on this?

mezz commented:
Looks good but I think the test mod is not working as intended.
From the TextureDump output, I just see a square outline instead of the Book and Pickaxe outlines.
![texture_atlas_dump_textures_mipmap_0](https://cloud.githubusercontent.com/assets/916092/25575631/f3e0fe62-2e0d-11e7-8f41-9da49cce6252.png)

Also the test mod should use the same formatting as the rest of Forge, with braces on new lines.

marchermans commented:
Hmm The pictures above were taken from the texture dump generated by the same test mod
Let me Check what is going on. I will update the formatting as well.

marchermans committed:
Updating the Formatting.

Based on TiCon and Armory Code.

marchermans commented:
I ran the tests again, and i am getting the same result as above.
I am trying to figure out what is happening.

marchermans commented:
@mezz How are you running the test?

liach commented:
Eh, the test look extremely complex to me.

marchermans commented:
@liach it finds the edge of a texture by iterating over all its pixels.
It is not that complicated.

mezz commented:
I use IntelliJ to run the test mods within the Forge dev environment.
Here's a video on how to set it up: https://www.youtube.com/watch?v=pLWQk6ed56Q

The trick is to build the Tests project so it creates all the classes, and then set the classes folder as a runtime dependency for the Main project. 

![](http://i.imgur.com/VZj5zPC.png)

That way the test will be run when you run the main project, and to update them you just have to build the Test project.

marchermans commented:
Ah okey
I will give that a try, i usually create a dummy copy in the main module.

mezz commented:
I pulled the changes and it works now.
In the diff it looks like you replaced the two calls to `getAtlasSprite` with `getTextureExtry`, maybe that fixed it and you didn't commit the change before.

mezz commented:
Tested and looks good.

**TLDR** for Lex (sorry Orion but you are extremely verbose!):
This event allows mods to create textures based on other textures in the atlas.
The example mod takes the book and pickaxe texture and creates a new texture that's just their outline.
![](https://cloud.githubusercontent.com/assets/5585406/24015762/ae3755f2-0a89-11e7-83dd-662dd97e880b.png)

RainWarrior commented:
What happens if someone wants to depend on one of the textures added in this event? In the [example](https://github.com/SmithsGaming/SmithsCore/blob/Development-1.11/src/com/smithsmodding/smithscore/client/textures/TextureCreator.java#L93), you're using EventPriority.LOWEST, which, from my understanding, means that nothing else will ba able to run after you. Wouldn't a system with an explicit dependencies be better? An event solution like this doesn't actually solve the problem of texture dependencies, it postpones it. Unless I've missed something :)

mezz commented:
What you linked to and are discussing is not the example used in this PR.

marchermans commented:
@Mezz what do you mean?

mezz commented:
Fry was looking at the old implementation and not the one in the PR. I talked to him on IRC and his argument still stands.

marchermans committed:
Changed the Implementation: System now allows dependence on textures that already depend on an existing texture.

marchermans commented:
@mezz @RainWarrior I changed it. I tested my test mod again and this works in both cases.
Also to display that event works for textures that depend on a other texture i added log output that shows you that the events gets fired properly for textures i register while i am in the event handler.

marchermans committed:
Fixing the imports in ForgeHooksClient

tterrag1098 commented:
@OrionDevelopment Thank you for this, the earlier PR breaks my json system as well, since I need to parse texture metadata for additional textures to stitch. So one more 👍 from me.

Baylem commented:
@OrionDevelopment I love this, the previous PR by liach messed up my systems as well. so a :+1: from me

liach commented:
To all of you @tterrag1098 @OhCreative, if I added one event back then (mezz actually suggested to mess this up, see comments) things would be fine!

tterrag1098 commented:
@liach I think these two PRs should have happened at once, there is good reason to have both phases, but clearly there are issues (which this PR solves) with texture dependencies.

marchermans commented:
Sadly enough I am still waiting on it.

marchermans commented:
@liach Things like this can happen, no way you could have for seen somebody actually using it to make texture dependencies. If this gets pulled then it is all forgiven and forgotten :D

liach commented:
Waiting for fry. He seems to be away.

mezz commented:
#4070 has been accepted and merged, any further improvements are welcome. 
I think it means that this PR can be closed, is that right?

marchermans commented:
Yes i think so.
